### PR TITLE
iOS implementation and sample project

### PR DESCRIPTION
**Pull request by CherryPerry**

### CODE REVIEW

File: reaktive/src/iosCommonMain/kotlin/com/badoo/reaktive/scheduler/MainScheduler.kt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import com.badoo.reaktive.disposable.CompositeDisposable
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.utils.freeze
+import platform.darwin.DISPATCH_TIME_NOW
+import platform.darwin.dispatch_after
+import platform.darwin.dispatch_get_main_queue
+import platform.darwin.dispatch_time
+import kotlin.native.concurrent.AtomicReference
+
+class MainScheduler : Scheduler {
```
Comment (arkivanov) [inline reaktive/src/iosCommonMain/kotlin/com/badoo/reaktive/scheduler/MainScheduler.kt]: internal

File: reaktive/src/iosCommonMain/kotlin/com/badoo/reaktive/scheduler/MainScheduler.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+        override fun dispose() {
+            operations.dispose()
+        }
+
+        private class Operation(
+            task: () -> Unit
+        ) : () -> Unit, Disposable {
+
+            private val taskReference = AtomicReference<(() -> Unit)?>(task.freeze())
```
Comment (arkivanov) [inline reaktive/src/iosCommonMain/kotlin/com/badoo/reaktive/scheduler/MainScheduler.kt]: You can pass second parameter as "true", then this object (any following objects) will be frozen automatically.

File: reaktive/src/iosCommonMain/kotlin/com/badoo/reaktive/scheduler/MainScheduler.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+            override fun invoke() {
+                taskReference.value?.invoke()
+            }
+
+            override val isDisposed: Boolean
+                get() = taskReference.value == null
+
+            override fun dispose() {
+                taskReference.value = null
```
Comment (arkivanov) [inline reaktive/src/iosCommonMain/kotlin/com/badoo/reaktive/scheduler/MainScheduler.kt]: We should cancel tasks somehow, at least pending tasks

File: sample-ios-app/src/iosCommonMain/kotlin/com.badoo.reaktive.sample.ios/Timer.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import com.badoo.reaktive.single.Single
+import platform.Foundation.NSOperationQueue
+
+fun calculateFibonacci(): Single<List<Pair<Int, Int>>> =
+    observableOf(37, 38, 39, 40, 41)
+        .subscribeOn(computationScheduler)
+        .doOnBeforeNext { println("Should be background thread: ${!isMainThread()}") }
+        .map {
+            // some off-thread computations
+            Pair(it, fibonacci(it))
```
Comment (arkivanov) [inline sample-ios-app/src/iosCommonMain/kotlin/com.badoo.reaktive.sample.ios/Timer.kt]: Use "to" infix function

File: sample-ios-app/src/iosCommonMain/kotlin/com.badoo.reaktive.sample.ios/Timer.kt [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        }
+        .observeOn(mainScheduler)
+        .doOnBeforeNext { println("Should be main thread: ${isMainThread()}") }
+        .toList()
+
+private fun fibonacci(n: Int): Int =
+    when (n) {
+        0 -> 0
+        1 -> 1
+        else -> fibonacci(n - 1) + fibonacci(n - 2)
```
Comment (arkivanov) [inline sample-ios-app/src/iosCommonMain/kotlin/com.badoo.reaktive.sample.ios/Timer.kt]: C'mon, use dynamic programming :-D

File: sample-ios-app/xcode/.gitignore [hunk 6]
```diff
@@ -0,0 +1,18 @@
+build/
```
Comment (arkivanov) [inline sample-ios-app/xcode/.gitignore]: Perhaps can be omitted

File: reaktive/src/iosCommonMain/kotlin/com/badoo/reaktive/scheduler/MainScheduler.kt (hunk 3 above)
Comment (CherryPerry) [reply reaktive/src/iosCommonMain/kotlin/com/badoo/reaktive/scheduler/MainScheduler.kt]: `cancel()` == `dispose()`. When operation is disposed, task reference is `null` in `invoke()`.

Comment (arkivanov) [reply reaktive/src/iosCommonMain/kotlin/com/badoo/reaktive/scheduler/MainScheduler.kt]: But about pending tasks? That are in the queue (delayed and/or repeating).

File: sample-ios-app/src/iosCommonMain/kotlin/com.badoo.reaktive.sample.ios/Timer.kt (hunk 5 above)
Comment (CherryPerry) [reply sample-ios-app/src/iosCommonMain/kotlin/com.badoo.reaktive.sample.ios/Timer.kt]: I don't want to use dynamic programming, I want this calculation be slow and executed on background thread. Replaced with `sleep`.

File: reaktive/src/iosCommonMain/kotlin/com/badoo/reaktive/scheduler/MainScheduler.kt (hunk 3 above)
Comment (CherryPerry) [reply reaktive/src/iosCommonMain/kotlin/com/badoo/reaktive/scheduler/MainScheduler.kt]: `cancel()` calls `disposables.clear()`. Each operation is `Disposable` and is added into disposables collection.

Comment (CherryPerry) [reply reaktive/src/iosCommonMain/kotlin/com/badoo/reaktive/scheduler/MainScheduler.kt]: If task is in pending state (not executed yet), then reference to task will become `null`. So dispatched operation becomes no-op. If you are talking about some system `remove_dispatch(...)` method, then there is no such method, we can only add operations to queue.

**Pull request merged**